src/clients/python/experimental_api_v2/examples/v2_image_client.py

```diff
@@ -373,20 +408,73 @@ def requestGenerator(supports_batching, input_name, output_name, c, h, w, format
     # start over with the first images until the batch is filled.
     requests = []
     responses = []
+    user_data = UserData()
+
+    # Used to hold the greenlets running the intference in HTTP
+    # async operation.
+    greenlets = []
+
+    sent_count = 0
 
     # Send request
     try:
+        if FLAGS.streaming:
+            triton_client.start_stream(partial(completion_callback, user_data))
+
         for inputs, outputs, model_name, model_version in requestGenerator(
-                max_batch_size > 0, input_name, output_name, c, h, w, format, dtype, FLAGS):
-            responses.append(
-                triton_client.infer(FLAGS.model_name,
-                                    inputs,
-                                    model_version=FLAGS.model_version,
-                                    outputs=outputs))
+                max_batch_size > 0, input_name, output_name, c, h, w, format,
+                dtype, FLAGS):
+            sent_count += 1
+            if FLAGS.streaming:
+                triton_client.async_stream_infer(
+                    FLAGS.model_name,
+                    inputs,
+                    model_version=FLAGS.model_version,
+                    outputs=outputs)
+            elif FLAGS.async_set:
+                if FLAGS.protocol.lower() == "grpc":
+                    triton_client.async_infer(FLAGS.model_name,
+                                              inputs,
+                                              partial(completion_callback,
+                                                      user_data),
+                                              model_version=FLAGS.model_version,
+                                              outputs=outputs)
+                else:
+                    greenlets.append(
+                        triton_client.async_infer(
+                            FLAGS.model_name,
+                            inputs,
+                            partial(completion_callback, user_data),
+                            model_version=FLAGS.model_version,
+                            outputs=outputs))
+            else:
+                responses.append(
+                    triton_client.infer(FLAGS.model_name,
+                                        inputs,
+                                        model_version=FLAGS.model_version,
+                                        outputs=outputs))
 
     except InferenceServerException as e:
         print("inference failed: " + str(e))
-        sys.exit()
+        if FLAGS.streaming:
+            triton_client.stop_stream()
+        sys.exit(1)
+
```

Comment:
I'm curious, what would the behavior be if the program exits before stop_stream is called?

Reply:
The stream will get stuck and system will hang. Python, unfortunately, is not very adept at cleanups(destructor logic). If you look at our library we use an iterator class to keep the stream up and running. The order of clean up is not guaranteed by the python's garbage collector. We can only clean the client object when the stream itself is closed, which happens when None object is passed into the stream. If the None object is not passed, then the stream is kept open preventing client from getting deleted.

Reply:
"We can only clean the client object when the stream itself is closed", so the client's __del__ function will never be called unless the stream is closed? Otherwise, can we call stop_stream() in the __del__()?

Reply:
We have kept stop_stream in the del(). But the issue is python garbage collector will not call the del() until client ref is zero. While sending the requests on the grpc stream, grpc library copies this ref hence incrementing the counter. This leads to deadlock. Client's del() will not be called unless the stream is closed, and stream will not be closed unless None is passed by the stop_stream(). 

To avoid this precarious situation, it is best for user to ensure the stop_stream() is explicitly called.